common/lib/xmodule/xmodule/modulestore/tests/test_mixed_modulestore.py

```diff
@@ -2091,3 +2096,123 @@ def test_import_edit_import(self, default):
+
+                        self.store.publish(block.location, self.user_id)
+                        self.assertEqual(receiver.call_count, 3)
+
+                    # Test a draftable block type. Although blocks of this type are only published when the publish
+                    # method is explicitly called, our implementation will still emit the signal after each update.
+                    receiver.reset_mock()
+                    block = self.store.create_item(self.user_id, course_key, 'problem')
+                    self.store.update_item(block, self.user_id)
+                    self.assertEqual(receiver.call_count, 1)
+
+                    self.store.publish(block.location, self.user_id)
+                    self.assertEqual(receiver.call_count, 2)
+
+                    # Test course re-runs
+                    receiver.reset_mock()
+                    dest_course_key = self.store.make_course_key("org.other", "course.other", "run.other")
+                    self.store.clone_course(course_key, dest_course_key, self.user_id)
+                    self.assertEqual(receiver.call_count, 1)
+
+                    # Test course imports
+                    receiver.reset_mock()
+                    import_from_xml(
+                        self.store, self.user_id, DATA_DIR, ['toy'], load_error_modules=False,
+                        static_content_store=contentstore,
+                        create_course_if_not_present=True,
+                    )
+
+                    # Note: The signal is fired once when the course is created and
+                    # a second time after the actual data import.
+                    self.assertEqual(receiver.call_count, 2)
+
+    def test_course_publish_signal_firing_mongo(self):
+        with MongoContentstoreBuilder().build() as contentstore:
+            self.store = MixedModuleStore(
+                contentstore=contentstore,
+                create_modulestore_instance=create_modulestore_instance,
+                mappings={},
+                signal_handler=SignalHandler(MixedModuleStore),
+                **self.OPTIONS
+            )
+            self.addCleanup(self.store.close_all_connections)
+
+            with self.store.default_store(ModuleStoreEnum.Type.mongo):
+                self.assertIsNotNone(self.store.thread_cache.default_store.signal_handler)
+
+                with mock_signal_receiver(SignalHandler.course_published) as receiver:
+                    self.assertEqual(receiver.call_count, 0)
+
+                    # Course creation and publication should fire the signal
+                    course = self.store.create_course('org_x', 'course_y', 'run_z', self.user_id)
+                    self.store.publish(course.location, self.user_id)
+                    self.assertEqual(receiver.call_count, 1)
+
+                    course_key = course.id
+
+                    # The signal should only be fired if the publish method is explicitly called.
```

Comment:
This means changes to blocks whose type is in `DIRECT_ONLY_CATEGORIES` will not fire a signal - even though the published version of those blocks are updated.  This will cause the cached API result to remain invalid.  

To fix this, you can update `update_item` in old Mongo and signal if 
1. the update is happening on a "published" version of the block, and
2. there is no bulk_operation in progress.